plugins/tech-insights-backend/src/service/persistence/TechInsightsDatabase.ts

```diff
@@ -0,0 +1,177 @@
+  TechInsightsStore,
+} from '@backstage/plugin-tech-insights-common';
+import { rsort } from 'semver';
+import { groupBy } from 'lodash';
+import { DateTime } from 'luxon';
+
+export type RawDbFactRow = {
+  ref: string;
+  version: string;
+  timestamp: string;
+  entity: string;
+  facts: string;
+};
+
+type RawDbFactSchemaRow = {
+  id: number;
+  ref: string;
+  version: string;
+  schema: string;
+};
+
+const migrationsDir = resolvePackagePath(
+  '@backstage/plugin-tech-insights-backend',
+  'migrations',
+);
+
+export class TechInsightsDatabase implements TechInsightsStore {
+  private readonly CHUNK_SIZE = 50;
+  static async create(knex: Knex): Promise<TechInsightsStore> {
+    await knex.migrate.latest({
+      directory: migrationsDir,
+    });
+    return new TechInsightsDatabase(knex);
+  }
+
+  constructor(private readonly db: Knex) {}
+
+  private async getLatestSchema(ref: string): Promise<RawDbFactSchemaRow> {
+    const existingSchemas = await this.db<RawDbFactSchemaRow>('fact_schemas')
+      .where({ ref })
+      .orderBy('id', 'desc')
+      .select();
+    if (existingSchemas.length < 1) {
+      throw new Error(`No schema found for ${ref}. `);
+    }
+    const sorted = rsort(existingSchemas.map(it => it.version));
+    return existingSchemas.find(it => it.version === sorted[0])!!;
+  }
+
+  async getLatestSchemas(refs?: string[]): Promise<FactSchema[]> {
+    const queryBuilder = this.db<RawDbFactSchemaRow>('fact_schemas');
+    if (refs) {
+      queryBuilder.whereIn('ref', refs);
+    }
+    const existingSchemas = await queryBuilder.orderBy('id', 'desc').select();
+
+    const groupedSchemas = groupBy(existingSchemas, 'ref');
+    return Object.values(groupedSchemas)
+      .map(schemas => {
+        const sorted = rsort(schemas.map(it => it.version));
+        return schemas.find(it => it.version === sorted[0])!!;
+      })
+      .map((it: RawDbFactSchemaRow) => ({
+        ...it,
+        schema: JSON.parse(it.schema),
+      }));
+  }
+
+  async insertFactSchema(ref: string, schema: FactSchema) {
+    const existingSchemas = await this.db<RawDbFactSchemaRow>('fact_schemas')
+      .where({ ref })
+      .select();
+    const exists = existingSchemas.some(
+      it => it.ref === ref && it.version === schema.version,
+    );
+
+    if (!exists) {
+      await this.db<RawDbFactSchemaRow>('fact_schemas').insert({
+        ref,
+        version: schema.version,
+        schema: JSON.stringify(schema.schema),
+      });
+    }
+  }
+
+  async insertFacts(facts: TechInsightFact[]): Promise<void> {
+    const tx = await this.db.transaction();
+    if (facts.length === 0) return;
+    const currentSchema = await this.getLatestSchema(facts[0].ref);
+    const factRows = facts.map(it => {
+      const { namespace, name, kind } = it.entity;
+      return {
+        ref: it.ref,
+        version: currentSchema.version,
+        entity: `${namespace.toLowerCase()}/${kind.toLowerCase()}/${name.toLowerCase()}`,
```

Comment:
you could just call toLowerCase once on the interpolated result ? (super trivial but I couldn't help it)